src/psbt.c

```diff
@@ -712,825 +801,1074 @@ int wally_psbt_init_alloc(
     size_t global_unknowns_allocation_len,
     struct wally_psbt **output)
 {
-    struct wally_psbt_input *new_inputs = NULL;
-    struct wally_psbt_output *new_outputs = NULL;
     struct wally_psbt *result;
+    int ret;
 
     TX_CHECK_OUTPUT;
     TX_OUTPUT_ALLOC(struct wally_psbt);
 
-    if (inputs_allocation_len) {
-        new_inputs = wally_malloc(inputs_allocation_len * sizeof(struct wally_psbt_input));
-        wally_bzero(new_inputs, inputs_allocation_len * sizeof(*new_inputs));
-    }
-    if (outputs_allocation_len) {
-        new_outputs = wally_malloc(outputs_allocation_len * sizeof(struct wally_psbt_output));
-        wally_bzero(new_outputs, outputs_allocation_len * sizeof(*new_outputs));
-    }
-    wally_unknowns_map_init_alloc(global_unknowns_allocation_len, &result->unknowns);
-    if ((inputs_allocation_len && !new_inputs) ||
-        (outputs_allocation_len && !new_outputs) ||
-        (global_unknowns_allocation_len && !result->unknowns)) {
-        wally_free(new_inputs);
-        wally_free(new_outputs);
-        wally_free(result->unknowns);
-        wally_free(result);
-        *output = NULL;
-        return WALLY_ENOMEM;
+    if (inputs_allocation_len)
+        result->inputs = wally_calloc(inputs_allocation_len * sizeof(struct wally_psbt_input));
+    if (outputs_allocation_len)
+        result->outputs = wally_calloc(outputs_allocation_len * sizeof(struct wally_psbt_output));
+
+    ret = wally_unknowns_map_init_alloc(global_unknowns_allocation_len, &result->unknowns);
+
+    if (ret != WALLY_OK ||
+        (inputs_allocation_len && !result->inputs) || (outputs_allocation_len && !result->outputs)) {
+        wally_psbt_free(result);
+        return ret != WALLY_OK ? ret : WALLY_ENOMEM;
     }
 
-    result->inputs = new_inputs;
-    result->num_inputs = 0;
+    /* Version is always 0 */
```

Comment:
`wally_tx_init_alloc` allows to set the transaction version, should we allow the user to set the psbt version too in `wally_psbt_init_alloc` and its elements variant? It could be needed in the future.

Reply:
Yes, good idea - pushed.